trollflow2/launcher.py

```diff
@@ -115,40 +116,82 @@ def check_results(produced_files, start_time, exitcode):
         LOG.info('All files produced nominally in %s.', str(elapsed), extra={'time': elapsed})
 
 
-def run(prod_list, topics=None, test_message=None, nameserver='localhost',
-        addresses=None):
-    """Spawn one or multiple subprocesses to run the jobs from the product list."""
-    LOG.info("Launching trollflow2")
-    tmessage = get_test_message(test_message)
-    if tmessage:
-        from threading import Thread as Process
-        from six.moves.queue import Queue
-        from posttroll.message import Message
-    else:
-        from multiprocessing import Process, Queue
-
-    with open(prod_list) as fid:
-        config = yaml.load(fid.read(), Loader=BaseLoader)
-    topics = topics or config['product_list'].pop('subscribe_topics', None)
-
-    if not tmessage:
-        listener = ListenerContainer(topics=topics, nameserver=nameserver,
-                                     addresses=addresses)
-
+def generate_messages(connection_parameters):
+    """Generate messages using a ListenerContainer."""
+    listener = _create_listener_from_connection_parameters(connection_parameters)
     while True:
         try:
-            if tmessage:
-                msg = Message(rawstr=tmessage)
-            else:
-                msg = listener.output_queue.get(True, 5)
+            yield listener.output_queue.get(True, 5)
         except KeyboardInterrupt:
-            if not tmessage:
-                listener.stop()
+            listener.stop()
             return
         except Empty:
             continue
+
+
+def _create_listener_from_connection_parameters(connection_parameters):
+    """Create listener from connection parameters."""
+    topics = connection_parameters['topic']
+    nameserver = connection_parameters.get('nameserver', 'localhost')
+    addresses = connection_parameters.get('addresses')
+    listener = ListenerContainer(
+        addresses=addresses,
+        nameserver=nameserver,
+        topics=topics)
+    return listener
+
+
+def run(product_list, connection_parameters=None, test_message=None):
+    """Spawn one or multiple subprocesses to run the jobs from the product list."""
+    test_message = get_test_message(test_message)
+    if test_message:
+        return _run_threaded(product_list, test_message)
+    return _run_subprocess(product_list, connection_parameters)
```

Comment:
Neither of these should return anything.

```suggestion
    if test_message:
        _run_threaded(product_list, test_message)
    else:
        _run_subprocess(product_list, connection_parameters)
```